File: C++/25.reverse-nodes-in-k-group.cpp
/*
 * @lc app=leetcode.cn id=25 lang=cpp
 *
 * [25] Reverse Nodes in k-Group
 */

/**
 * Definition for singly-linked list.
 */
#include<iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
// @lc code=start
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int count = 0;
        ListNode *pre = head;
        while(pre) {
            ++count;
            pre = pre->next;
        }
        if (k<=1 || count<k)
            return head;

        ListNode *cur = head, *next=NULL;
        pre = NULL;
        for (int i=0;i<k;++i) {
            next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        } 
        head->next = reverseKGroup(cur, k);
        return pre;
    }
};
// @lc code=end

File: Python/25.reverse-nodes-in-k-group.py
#
# @lc app=leetcode id=25 lang=python
#
# [25] Reverse Nodes in k-Group
#
# https://leetcode.com/problems/reverse-nodes-in-k-group/description/
#
# algorithms
# Hard (38.39%)
# Likes:    1492
# Dislikes: 299
# Total Accepted:    216.5K
# Total Submissions: 562.1K
# Testcase Example:  '[1,2,3,4,5]\n2'
#
# Given a linked list, reverse the nodes of a linked list k at a time and
# return its modified list.
# 
# k is a positive integer and is less than or equal to the length of the linked
# list. If the number of nodes is not a multiple of k then left-out nodes in
# the end should remain as it is.
# 
# 
# 
# 
# Example:
# 
# Given this linked list: 1->2->3->4->5
# 
# For k = 2, you should return: 2->1->4->3->5
# 
# For k = 3, you should return: 3->2->1->4->5
# 
# Note:
# 
# 
# Only constant extra memory is allowed.
# You may not alter the values in the list's nodes, only nodes itself may be
# changed.
# 
# 
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """

        if not head or k == 1:
            return head

        cur = head
        for _ in range(k):
            if not cur:
                return head
            cur = cur.next

        cur = head
        pre = None
        for _ in range(k):
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt

        head.next = self.reverseKGroup(cur, k)
        return pre

# @lc code=end

